yesod-auth/Yesod/Auth/Email.hs

```diff
@@ -753,14 +771,20 @@ postPasswordR = do
                               then getThird jcreds
                               else fcurrent
                 mrealpass <- lift $ getPassword aid
-                case mrealpass of
-                    Nothing ->
+                case (mrealpass, current) of
+                    (Nothing, _) ->
                         lift $ loginErrorMessage (tm setpassR) "You do not currently have a password set on your account"
-                    Just realpass
-                        | isNothing current -> loginErrorMessageI LoginR Msg.BadSetPass
-                        | isValidPass (fromJust current) realpass -> confirmPassword aid tm jcreds
-                        | otherwise ->
-                            lift $ loginErrorMessage (tm setpassR) "Invalid current password, please try again"
+                    (_, Nothing) ->
+                        lift $ loginErrorMessage (tm loginR) "You did not submit a password, please try again"
+                    (Just realpass, Just current') -> do
+                        passValid <- lift $ verifyPassword current' realpass
+                        case () of
+                            _
+                                | isNothing current -> loginErrorMessageI LoginR Msg.BadSetPass
```

Comment:
Can this check be avoided as we are doing this on top .also ?

Reply:
ahhh great point. in fact, this is the proper error message to deliver. refactoring & updating version